src/rdsn/src/meta/meta_backup_engine.cpp

```diff
@@ -86,87 +86,57 @@ void meta_backup_engine::start()
-void meta_backup_engine::update_backup_item_on_remote_storage(backup_status::type new_status,
-                                                              int64_t end_time)
-{
-    // TODO(heyuchen): TBD
-}
-
-// TODO(heyuchen): update following functions
-
-error_code meta_backup_engine::write_backup_file(const std::string &file_name,
-                                                 const dsn::blob &write_buffer)
-{
-    dist::block_service::create_file_request create_file_req;
-    create_file_req.ignore_metadata = true;
-    create_file_req.file_name = file_name;
-
-    dsn::error_code err;
-    dist::block_service::block_file_ptr remote_file;
-    _block_service
-        ->create_file(create_file_req,
-                      TASK_CODE_EXEC_INLINED,
-                      [&err, &remote_file](const dist::block_service::create_file_response &resp) {
-                          err = resp.err;
-                          remote_file = resp.file_handle;
-                      })
-        ->wait();
-    if (err != dsn::ERR_OK) {
-        ddebug_f("create file {} failed", file_name);
-        return err;
-    }
-    dassert_f(remote_file != nullptr,
-              "create file {} succeed, but can't get handle",
-              create_file_req.file_name);
-    remote_file
-        ->write(dist::block_service::write_request{write_buffer},
-                TASK_CODE_EXEC_INLINED,
-                [&err](const dist::block_service::write_response &resp) { err = resp.err; })
-        ->wait();
-    return err;
-}
-
-error_code meta_backup_engine::backup_app_meta()
-{
-    dsn::blob app_info_buffer;
+    blob app_info_buffer;
     {
         zauto_read_lock l;
-        _backup_service->get_state()->lock_read(l);
-        std::shared_ptr<app_state> app = _backup_service->get_state()->get_app(_cur_backup.app_id);
+        _meta_svc->get_server_state()->lock_read(l);
+        std::shared_ptr<app_state> app = _meta_svc->get_server_state()->get_app(_cur_backup.app_id);
         if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
             derror_f("app {} is not available, couldn't do backup now.", _cur_backup.app_id);
             return ERR_INVALID_STATE;
         }
         app_state tmp = *app;
-        // Because we don't restore app envs, so no need to write app envs to backup file.
-        // TODO(zhangyifan): backup and restore app envs when needed.
-        tmp.envs.clear();
-        app_info_buffer = dsn::json::json_forwarder<app_info>::encode(tmp);
+        app_info_buffer = json::json_forwarder<app_info>::encode(tmp);
     }
 
-    std::string backup_root =
-        dsn::utils::filesystem::path_combine(_backup_path, _backup_service->backup_root());
-    // TODO(heyuchen): refactor and update it in future
-    std::string file_name = "todo";
-    return write_backup_file(file_name, app_info_buffer);
+    std::string remote_meta_dir =
+        get_backup_meta_path(get_backup_root(FLAGS_cold_backup_root, _cur_backup.backup_path),
```

Comment:
can we move `get_backup_root` into `get_backup_meta_path`, and only left `_cur_backup.backup_path` here?